oss-internship-2020/libtiff/test/raw_decode.cc

```diff
@@ -0,0 +1,244 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <array>
+#include <cstring>
+
+#include "helper.h"  // NOLINT(build/include)
+#include "tiffio.h"  // NOLINT(build/include)
+
+namespace {
+
+using ::sapi::IsOk;
+using ::testing::Eq;
+using ::testing::IsFalse;
+using ::testing::IsTrue;
+using ::testing::NotNull;
+
+struct ChannelLimits {
+  uint8_t min_red;
+  uint8_t max_red;
+  uint8_t min_green;
+  uint8_t max_green;
+  uint8_t min_blue;
+  uint8_t max_blue;
+  uint8_t min_alpha;
+  uint8_t max_alpha;
+};
+
+constexpr uint32_t kRawTileNumber = 9;
+constexpr uint32_t kClusterSize = 6;
+constexpr uint32_t kChannelsInPixel = 3;
+constexpr uint32_t kTestCount = 3;
+constexpr uint32_t kImageSize = 128 * 128;
+constexpr uint32_t kClusterImageSize = 64 * 64;
+using ClusterData = std::array<uint8_t, kClusterSize>;
+
+constexpr std::array<std::pair<uint32_t, ClusterData>, kTestCount> kClusters = {
+    {{0, {0, 0, 2, 0, 138, 139}},
+     {64, {0, 0, 9, 6, 134, 119}},
+     {128, {44, 40, 63, 59, 230, 95}}}};
+
+constexpr std::array<std::pair<uint32_t, ChannelLimits>, kTestCount> kLimits = {
+    {{0, {15, 18, 0, 0, 18, 41, 255, 255}},
+     {64, {0, 0, 0, 0, 0, 2, 255, 255}},
+     {512, {5, 6, 34, 36, 182, 196, 255, 255}}}};
+
+bool CheckCluster(uint32_t cluster, const sapi::v::Array<uint8_t>& buffer,
+                  const ClusterData& expected_cluster) {
+  bool is_overrun = (buffer.GetSize() <= cluster * kClusterSize);
+  EXPECT_THAT(is_overrun, IsFalse()) << "Overrun";
```

Comment:
This will not give a very useful error message.
```c++
EXPECT_THAT(buffer.GetSize(), Gt(cluster * kClusterSize));
```
will print the actual and expected values.